Initial prototype of a "slave" worker

erikbern opened the pull request:
If a worker is created using slave=True, then it will pull *any* tasks from the scheduler,
not just tasks that are assigned to it.

The biggest limitation atm is that the slave needs to be aware of what those tasks map to,
i.e. somehow they have to be imported for the worker to know what to do with the tasks.

It also doesn't handle failure very well.

Changed the semantics of how interface.load_task works a bit. It will now rely on the
Register class to locate task classes rather than importing modules dynamically. There is
still an option to import modules but it's not run by default.

erikbern commented:
Rebased this, getting even simpler. I think I can add the dynamic module loading pretty easy. Will keep experimenting.

My goal is to build the next generation of the scheduler by triggering jobs from the scheduler to a worker pool.

Also did anyone have any thoughts about what word to use instead of "slave"?

Tarrasch commented:
What about "volunteer"? They are in fact not forced into work, they volunteer and happily take the workload the server assigns to them, so that word would make more sense. Or is my understanding of this new worker type not correct?

daveFNbuck commented:
How about "dedicated worker" for a name?

I just noticed that some of my larger backfills have been silently failing lately because they use too much memory to fork all the workers and hadoop jobs successfully. Looking forward to this to mitigate the problem.

erikbern commented:
Maybe deputy :) or assistant, executor, clerk, contractor, laborer, workforce

erikbern commented:
I think it would be great if we can pick this up an start introducing more and more elements of a purely distributed task execution framework. Unfortunately my time is very constrained right now. But I can at least try to rebase this feature and we can merge it as an experimental feature

Tarrasch commented:
Sure. As long as it's optional and doesn't break anything. ^^

erikbern commented:
Of course

daveFNbuck commented:
If we can get it in as an experimental feature, I'll help iterate on it as a test subject.

erikbern commented:
That's awesome. I'll take a look at it.

erikbern commented:
Rebased it. I also renamed "slave" to "assistant"

There's an annoying circular dependency between `luigi.worker` and `luigi.interface` that I solved by using a dynamic import. We should refactor soon (both are huge).

gpoulin commented:
@erikbern This import loop also caused me issues when porting to python3 (e3ec5f46). Would be nice to refactor this.

erikbern commented:
Ok, I'll find some time to address it.

erikbern commented:
Does anyone object if I merge this? It's an undocumented feature (so far) but it's a starting point for a lot of cool things and it's awesome if someone (like @daveFNbuck or @tommyengstrom) has time to look at it

daveFNbuck commented:
Please do. I'll start using it today.
